### PR TITLE
Plugin messaging Docs

### PR DESCRIPTION
**Pull request by olijeffers0n**
This takes the content of #84 and supersedes it. That PR has been abandoned and this is a topic that comes up frequently. 

Please let me know if any of the technical details are incorrect :)

### CODE REVIEW

File: docs/paper/dev/api/plugin-messaging.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+# Plugin Messaging
+
+First introduced in [2012](https://web.archive.org/web/20220711204310/https://dinnerbone.com/blog/2012/01/13/minecraft-plugin-channels-messaging/),
+Plugin messaging is a way for Paper plugins to communicate with clients. When your servers are behind a proxy, 
+it will allow your Paper plugins to communicate with the proxy server.
+
+## Sending Plugin Messages
+
+First, we're actually going take a look at your Paper server. Your Paper server plugin will need to register that it 
+will be sending on any given plugin channel. You'll should to do this alongside your other event listener registrations.
```
Comment (kennytv) [inline docs/paper/dev/api/plugin-messaging.md:14]:
```suggestion
First, we're going to take a look at your Paper server. Your Paper server plugin will need to register that it 
will be sending on any given plugin channel. You should to do this alongside your other event listener registrations.
```

File: docs/paper/dev/api/plugin-messaging.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+is not connected to the server, the server will not be able to send or receive plugin messages.
+
+:::
+
+### What did we just do?
+
+We sent a plugin message on the `bungeecord:main` channel! The message we sent was a byte array that contained two strings converted to bytes: `Connect` and `hub2`.
+
+Our proxy server received the message through the player who triggered the `PlayerJumpEvent` on our Java server. 
+Then, it recognized the channel as it's own, and in alignment with BungeeCord's format sent our player to the `hub2` server.
```
Comment (kennytv) [inline docs/paper/dev/api/plugin-messaging.md:75]:
```suggestion
Then, it recognized the channel as its own and in alignment with BungeeCord's format sent our player to the `hub2` server.
```

File: docs/paper/dev/api/plugin-messaging.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+For BungeeCord, we can think of this message as a case-sensitive command with arguments. 
+Here, our command is `Connect` and our only argument is `hub2` but some "commands" may have multiple arguments. 
+For other channels introduced by client side mods, refer to their documentation to best understand how to format your messages.
+
+### BungeeCord Plugin Message Types
+
+Although we sent a `Connect` message to the proxy there are a few other cases that proxies will act on.
```
Comment (kennytv) [inline docs/paper/dev/api/plugin-messaging.md:83]:
```suggestion
Although we sent a `Connect` message to the proxy, there are a few other cases that proxies will act on. 
```

**Review by kennytv — APPROVED** [review]

File: docs/paper/dev/api/plugin-messaging.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        if (subchannel.equals("PlayerCount")) {
+            // This is our response to the PlayerCount request
+            String server = in.readUTF();
+            int playerCount = in.readInt();
+        }
+    }
+}
+```
+
+##### Example: `Forward`
```
Comment (lynxplay) [inline docs/paper/dev/api/plugin-messaging.md:137]: Would it be useful to at least throw some bare bone "what exactly are you trying to do" text here?
It obviously can be understood by *reading* the code below, however these docs are meant for developers learning this. It hence might be useful to give them a heads up as to what these examples try to accomplish.

**Review by the456gamer — COMMENTED** [review]
1 thing i noticed while looking at the deploy preview

File: docs/paper/dev/api/plugin-messaging.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        getServer().getMessenger().registerOutgoingPluginChannel(this, "bungeecord:main");
+        // Blah blah blah...
+    }
+
+}
+```
+
+Now that we're registered, we can send messages on the `bungeecord:main` channel.
+
+:::tip[The "bungeecord" and "bungeecord:main" channel]
```
[the remark this answers is not in the document]
Comment (olijeffers0n) [reply docs/paper/dev/api/plugin-messaging.md:30]: Yikes, that seems like a bit of a bug on docosaurus’ part tbh. I’d imagine it’s cause it’s looking for the colons to end the admonition, but thanks!

**Comment by olijeffers0n** [comment]
Added a note to the forward example

File: docs/velocity/dev/how-to/plugin-messaging.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+By default, your Velocity server will respond to the `bungeecord:main` channel unless you have disabled 
+`bungee-plugin-message-channel` in the Velocity configuration. Let's take a look at how plugin messaging works, and more importantly, how your Velocity server is already prepared to handle it.
+
+## Sending Plugin Messages
+
+Firstly, our Velocity server will need to register that it will be sending on any given plugin channel. 
+You should to do this alongside your other event listener registrations:
+
+```java
+
```
Comment (lynxplay) [inline docs/velocity/dev/how-to/plugin-messaging.md:22]: Useless empty line.

File: docs/paper/dev/api/plugin-messaging.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    }
+}
+```
+
+##### Example: `Forward`
+
+Forwards a plugin message to another server. This is useful for server-to-server communication within a proxy network.
+For example, if a certain player is banned on one server, you can forward a message to all other servers to ban them there too.
+
+:::caution[Example of banning a player on all servers]
```
Comment (lynxplay) [inline docs/paper/dev/api/plugin-messaging.md:144]: I feel like, we'd wanna move this like, behind the following example.
Reading this had me think that the following example was for banning players.

File: docs/velocity/dev/how-to/plugin-messaging.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
+---
+slug: /dev/plugin-messaging
+---
+
+# Plugin Messaging
+
+First introduced in [2012](https://web.archive.org/web/20220711204310/https://dinnerbone.com/blog/2012/01/13/minecraft-plugin-channels-messaging/),
```
Comment (lynxplay) [inline docs/velocity/dev/how-to/plugin-messaging.md:7]: Do we want to mention somewhere that it is bungeecord:main on the proxy side but BungeeCord on the server side ?

**Review by lynxplay — CHANGES_REQUESTED** [review]

**Review by lynxplay — APPROVED** [review]

File: docs/paper/dev/api/plugin-messaging.md [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+| Message Type      | Description                                            | Arguments                                                        | Response                                          |
+|:------------------|:-------------------------------------------------------|:-----------------------------------------------------------------|:--------------------------------------------------|
+| `Connect`         | Connects the player to the specified server.           | `server name`                                                    | N/A                                               |
+| `ConnectOther`    | Connects another player to the specified server.       | `player name`, `server name`                                     | N/A                                               |
+| `IP`              | Returns the IP of the specified player.                | `player name`                                                    | `IP`, `port`                                      |
+| `PlayerCount`     | Returns the number of players on the specified server. | `server name`                                                    | `server name`, player count`                      |
```
Comment (Lulu13022002) [inline docs/paper/dev/api/plugin-messaging.md:94]: The player count in the last column doesn't seems well formatted

**Pull request merged**